proposals/NNNN-symbol-visibility.md

```diff
@@ -0,0 +1,90 @@
+defined the possible linkages for names. This proposal updates how these
+linkages are represented in LLVM IR. The current implementation presents
+challenges for the SPIR-V backend due to inconsistencies with OpenCL. We propose
+that names with program linkage in HLSL should have external linkage and default
+visibility in LLVM IR, while names with external linkage in HLSL should have
+external linkage and hidden visibility in LLVM IR.
+
+## Motivation
+
+Consider the following HLSL snippet:
+
+```
+void external_linkage() { … }
+export void program_linkage() { … }
+```
+
+In llvm-ir, these function will be represented as:
+
+```
+define void @external_linkage()() local_unnamed_addr #0 !dbg !8 {
+  ret void, !dbg !12
+}
+
+define void @program_linkage()() local_unnamed_addr #1 !dbg !13 {
+  ret void, !dbg !14
+}
+
+attributes #0 = { ... } # no hlsl.export
+attributes #1 = { ... "hlsl.export" ...}
+```
+
+In the DirectX backend, there is a pass that will “finalize” the linkage. It
+will change `@external_linkage’s` linkage to internal, and remove the
+`hlsl.export` attribute from `@program_linkage`.
+
+The SPIR-V backend emits the `Export` linkage attribute for every symbol with
+external linkage. For the example above, `external_linkage` would be decorated
+with the `Export` linkage attribute giving it the equivalent of program linkage.
+We cannot change this without modifying the behaviour for OpenCL. OpenCL
+generates functions that look exactly like `external_linkage` and they require
+the `Export` linkage attribute in the SPIR-V.
+
+To be consistent with OpenCL, we must represent function with program linkage
+the way we currently represent functions with `external_linkage`. Then we can
+represent functions with external linkage as with some other attribute.
+
+## Proposed solution
+
+I propose mapping HLSL concepts found in section 3.6 of the HLSL specification
+as follows:
+
+HLSL concept                        | LLVM-IR concept
+:---------------------------------- | :---------------
+Translation unit                    | Translation unit
+Program (partially or fully linked) | Shared library
+
+Then, we can map the HLSL linkages to LLVM IR as follows:
+
+| HLSL Linkage                                                                                                   | LLVM-IR representation                                                                                                                                                                                                                                   |
+|----------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
```

Comment:
I won't always insist on the 80 column limit, but this is too wide. I recommended adding explanations of the HLSL linkage terms above. Perhaps they can be moved there and these could just reference them for more detail. That or the rationale can be listed in a following paragraph. Or both!

Reply:
Personally, I like information in the table. When looking at table it draws the eye, and makes it easy to find what information you want, and the connection.

However, I have trouble formatting tables with lots of content. The mark down formater I uses a syntax that GitHub does not understand to split a cell up over multiple lines. I'll leave it like this for now because I am not sure what to do with it.